# Medical Appointment No Shows Capstone

## 2. Data Wrangling

#### Data Collection

In [1]:
import os
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from datetime import datetime
%matplotlib inline

In [2]:
cwd = os.getcwd()
print(cwd)

/Users/asaadkhattab/Documents/DSProjects/MedicalAppointmentNoShowsCapstone


In [3]:
os.listdir(cwd)

['.DS_Store',
 'LICENSE',
 '2_DataWrangling.ipynb',
 'data2016.csv',
 'models',
 'README.md',
 '.gitignore',
 'figures',
 '.ipynb_checkpoints',
 '.git',
 'data']

#### Create SubFolders

In [4]:
newfolder = "/Users/asaadkhattab/Documents/DSProjects/MedicalAppointmentNoShowsCapstone/data"

try:
    os.mkdir(newfolder)
except OSError:
    print ("Creation of the directory %s failed" % newfolder)
else:
    print ("Successfully created the directory %s " % newfolder)

Creation of the directory /Users/asaadkhattab/Documents/DSProjects/MedicalAppointmentNoShowsCapstone/data failed


In [5]:
newfolder = "/Users/asaadkhattab/Documents/DSProjects/MedicalAppointmentNoShowsCapstone/figures"

try:
    os.mkdir(newfolder)
except OSError:
    print ("Creation of the directory %s failed" % newfolder)
else:
    print ("Successfully created the directory %s " % newfolder)

Creation of the directory /Users/asaadkhattab/Documents/DSProjects/MedicalAppointmentNoShowsCapstone/figures failed


In [6]:
newfolder = "/Users/asaadkhattab/Documents/DSProjects/MedicalAppointmentNoShowsCapstone/models"

try:
    os.mkdir(newfolder)
except OSError:
    print ("Creation of the directory %s failed" % newfolder)
else:
    print ("Successfully created the directory %s " % newfolder)

Creation of the directory /Users/asaadkhattab/Documents/DSProjects/MedicalAppointmentNoShowsCapstone/models failed


In [7]:
path = 'data2016.csv'
df = pd.read_csv(path, parse_dates=True, index_col=None)

In [8]:
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [9]:
df.tail(3)

PatientId  AppointmentID Gender          ScheduledDay  \
110524  1.557663e+13        5630692      F  2016-04-27T16:03:52Z   
110525  9.213493e+13        5630323      F  2016-04-27T15:09:23Z   
110526  3.775115e+14        5629448      F  2016-04-27T13:30:56Z   

              AppointmentDay  Age Neighbourhood  Scholarship  Hipertension  \
110524  2016-06-07T00:00:00Z   21   MARIA ORTIZ            0             0   
110525  2016-06-07T00:00:00Z   38   MARIA ORTIZ            0             0   
110526  2016-06-07T00:00:00Z   54   MARIA ORTIZ            0             0   

        Diabetes  Alcoholism  Handcap  SMS_received No-show  
110524         0           0        0             1      No  
110525         0           0        0             1      No  
110526         0           0        0             1      No

In [10]:
#Rename Columns in CSV.
df = df.rename(columns = {"AppointmentId":"AppointmentId", "Neighbourhood": "Neighborhood", "Hipertension": "Hypertension", 
                          "Handcap":"Handicap", "SMS_received":"SMSSent", 
                          "No-show":"NoShow"})

#### Data Definition

In [11]:
print(df.columns.values)

['PatientId' 'AppointmentID' 'Gender' 'ScheduledDay' 'AppointmentDay'
 'Age' 'Neighborhood' 'Scholarship' 'Hypertension' 'Diabetes' 'Alcoholism'
 'Handicap' 'SMSSent' 'NoShow']


In [12]:
df.shape

(110527, 14)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighborhood    110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hypertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handicap        110527 non-null  int64  
 12  SMSSent         110527 non-null  int64  
 13  NoShow          110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


In [14]:
df.count()

PatientId         110527
AppointmentID     110527
Gender            110527
ScheduledDay      110527
AppointmentDay    110527
Age               110527
Neighborhood      110527
Scholarship       110527
Hypertension      110527
Diabetes          110527
Alcoholism        110527
Handicap          110527
SMSSent           110527
NoShow            110527
dtype: int64

In [15]:
df.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921784e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hypertension       Diabetes     Alcoholism       Handicap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

             SMSSent  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [16]:
df.describe().T

count          mean           std           min  \
PatientId      110527.0  1.474963e+14  2.560949e+14  3.921784e+04   
AppointmentID  110527.0  5.675305e+06  7.129575e+04  5.030230e+06   
Age            110527.0  3.708887e+01  2.311020e+01 -1.000000e+00   
Scholarship    110527.0  9.826558e-02  2.976748e-01  0.000000e+00   
Hypertension   110527.0  1.972459e-01  3.979213e-01  0.000000e+00   
Diabetes       110527.0  7.186479e-02  2.582651e-01  0.000000e+00   
Alcoholism     110527.0  3.039981e-02  1.716856e-01  0.000000e+00   
Handicap       110527.0  2.224796e-02  1.615427e-01  0.000000e+00   
SMSSent        110527.0  3.210256e-01  4.668727e-01  0.000000e+00   

                        25%           50%           75%           max  
PatientId      4.172614e+12  3.173184e+13  9.439172e+13  9.999816e+14  
AppointmentID  5.640286e+06  5.680573e+06  5.725524e+06  5.790484e+06  
Age            1.800000e+01  3.700000e+01  5.500000e+01  1.150000e+02  
Scholarship    0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  
Hypertension   0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  
Diabetes       0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  
Alcoholism     0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  
Handicap       0.000000e+00  0.000000e+00  0.000000e+00  4.000000e+00  
SMSSent        0.000000e+00  0.000000e+00  1.000000e+00  1.000000e+00

In [17]:
print ( df.nunique() / df.shape[0] * 100 )

PatientId          56.365413
AppointmentID     100.000000
Gender              0.001810
ScheduledDay       93.686611
AppointmentDay      0.024428
Age                 0.094095
Neighborhood        0.073285
Scholarship         0.001810
Hypertension        0.001810
Diabetes            0.001810
Alcoholism          0.001810
Handicap            0.004524
SMSSent             0.001810
NoShow              0.001810
dtype: float64


In [18]:
df['PatientId'] = df['PatientId'].apply(lambda x: str(int(x)));
df['PatientId'] = pd.to_numeric(df['PatientId'])

In [19]:
df['Gender'] = df['Gender'].astype('category')

In [20]:
time_format_ScheduledDay = '%Y-%m-%d %H:%M:%S'
time_format_AppointmentDay = '%Y-%m-%d'

df['ScheduledDay'] = pd.to_datetime(df['ScheduledDay'], format=time_format_ScheduledDay) 
df['AppointmentDay'] = pd.to_datetime(df['AppointmentDay'], format=time_format_AppointmentDay).dt.date
df['AppointmentDay'] = pd.to_datetime(df['AppointmentDay'])

In [21]:
df.dtypes

PatientId                       int64
AppointmentID                   int64
Gender                       category
ScheduledDay      datetime64[ns, UTC]
AppointmentDay         datetime64[ns]
Age                             int64
Neighborhood                   object
Scholarship                     int64
Hypertension                    int64
Diabetes                        int64
Alcoholism                      int64
Handicap                        int64
SMSSent                         int64
NoShow                         object
dtype: object

In [22]:
df.isnull().values.any() 

False

In [23]:
df.sort_index(axis=1, ascending=True)

Age  Alcoholism AppointmentDay  AppointmentID  Diabetes Gender  \
0        62           0     2016-04-29        5642903         0      F   
1        56           0     2016-04-29        5642503         0      M   
2        62           0     2016-04-29        5642549         0      F   
3         8           0     2016-04-29        5642828         0      F   
4        56           0     2016-04-29        5642494         1      F   
...     ...         ...            ...            ...       ...    ...   
110522   56           0     2016-06-07        5651768         0      F   
110523   51           0     2016-06-07        5650093         0      F   
110524   21           0     2016-06-07        5630692         0      F   
110525   38           0     2016-06-07        5630323         0      F   
110526   54           0     2016-06-07        5629448         0      F   

        Handicap  Hypertension       Neighborhood NoShow        PatientId  \
0              0             1    JARDIM DA PENHA     No   29872499824296   
1              0             0    JARDIM DA PENHA     No  558997776694438   
2              0             0      MATA DA PRAIA     No    4262962299951   
3              0             0  PONTAL DE CAMBURI     No     867951213174   
4              0             1    JARDIM DA PENHA     No    8841186448183   
...          ...           ...                ...    ...              ...   
110522         0             0        MARIA ORTIZ     No    2572134369293   
110523         0             0        MARIA ORTIZ     No    3596266328735   
110524         0             0        MARIA ORTIZ     No   15576631729893   
110525         0             0        MARIA ORTIZ     No   92134931435557   
110526         0             0        MARIA ORTIZ     No  377511518121127   

        SMSSent              ScheduledDay  Scholarship  
0             0 2016-04-29 18:38:08+00:00            0  
1             0 2016-04-29 16:08:27+00:00            0  
2             0 2016-04-29 16:19:04+00:00            0  
3             0 2016-04-29 17:29:31+00:00            0  
4             0 2016-04-29 16:07:23+00:00            0  
...         ...                       ...          ...  
110522        1 2016-05-03 09:15:35+00:00            0  
110523        1 2016-05-03 07:27:33+00:00            0  
110524        1 2016-04-27 16:03:52+00:00            0  
110525        1 2016-04-27 15:09:23+00:00            0  
110526        1 2016-04-27 13:30:56+00:00            0  

[110527 rows x 14 columns]

In [24]:
df.head(3)

PatientId  AppointmentID Gender              ScheduledDay  \
0   29872499824296        5642903      F 2016-04-29 18:38:08+00:00   
1  558997776694438        5642503      M 2016-04-29 16:08:27+00:00   
2    4262962299951        5642549      F 2016-04-29 16:19:04+00:00   

  AppointmentDay  Age     Neighborhood  Scholarship  Hypertension  Diabetes  \
0     2016-04-29   62  JARDIM DA PENHA            0             1         0   
1     2016-04-29   56  JARDIM DA PENHA            0             0         0   
2     2016-04-29   62    MATA DA PRAIA            0             0         0   

   Alcoholism  Handicap  SMSSent NoShow  
0           0         0        0     No  
1           0         0        0     No  
2           0         0        0     No

In [25]:
range = df.aggregate([min, max]).T
print(range)

                                      min                        max
PatientId                           39217            999981631772427
AppointmentID                     5030230                    5790484
ScheduledDay    2015-11-10 07:13:56+00:00  2016-06-08 20:07:23+00:00
AppointmentDay        2016-04-29 00:00:00        2016-06-08 00:00:00
Age                                    -1                        115
Neighborhood                    AEROPORTO                 VILA RUBIM
Scholarship                             0                          1
Hypertension                            0                          1
Diabetes                                0                          1
Alcoholism                              0                          1
Handicap                                0                          4
SMSSent                                 0                          1
NoShow                                 No                        Yes


#### DATA CLEANING

The Patients' Ages are between -1 and 115. So I'm going to filter the dataframe between 0 and 100 years old.

In [26]:
df = df[(df.Age >= 0) & (df.Age <= 100)]

In [27]:
nh = df.Neighborhood.unique()
nh.sort()
print(nh)

['AEROPORTO' 'ANDORINHAS' 'ANTÔNIO HONÓRIO' 'ARIOVALDO FAVALESSA'
 'BARRO VERMELHO' 'BELA VISTA' 'BENTO FERREIRA' 'BOA VISTA' 'BONFIM'
 'CARATOÍRA' 'CENTRO' 'COMDUSA' 'CONQUISTA' 'CONSOLAÇÃO' 'CRUZAMENTO'
 'DA PENHA' 'DE LOURDES' 'DO CABRAL' 'DO MOSCOSO' 'DO QUADRO'
 'ENSEADA DO SUÁ' 'ESTRELINHA' 'FONTE GRANDE' 'FORTE SÃO JOÃO' 'FRADINHOS'
 'GOIABEIRAS' 'GRANDE VITÓRIA' 'GURIGICA' 'HORTO' 'ILHA DAS CAIEIRAS'
 'ILHA DE SANTA MARIA' 'ILHA DO BOI' 'ILHA DO FRADE' 'ILHA DO PRÍNCIPE'
 'ILHAS OCEÂNICAS DE TRINDADE' 'INHANGUETÁ' 'ITARARÉ' 'JABOUR'
 'JARDIM CAMBURI' 'JARDIM DA PENHA' 'JESUS DE NAZARETH' 'JOANA D´ARC'
 'JUCUTUQUARA' 'MARIA ORTIZ' 'MARUÍPE' 'MATA DA PRAIA' 'MONTE BELO'
 'MORADA DE CAMBURI' 'MÁRIO CYPRESTE' 'NAZARETH' 'NOVA PALESTINA'
 'PARQUE INDUSTRIAL' 'PARQUE MOSCOSO' 'PIEDADE' 'PONTAL DE CAMBURI'
 'PRAIA DO CANTO' 'PRAIA DO SUÁ' 'REDENÇÃO' 'REPÚBLICA' 'RESISTÊNCIA'
 'ROMÃO' 'SANTA CECÍLIA' 'SANTA CLARA' 'SANTA HELENA' 'SANTA LUÍZA'
 'SANTA LÚCIA' 'SANTA MARTHA' 'SANTA TEREZA

'ILHAS OCEÂNICAS DE TRINDADE' isn't a part of the neighborhood. 


In [28]:
df.drop(df[df['Neighborhood'] == 'ILHAS OCEÂNICAS DE TRINDADE'].index, inplace=True)

In [29]:
df = df.drop_duplicates()

In [30]:
df.count()

PatientId         110517
AppointmentID     110517
Gender            110517
ScheduledDay      110517
AppointmentDay    110517
Age               110517
Neighborhood      110517
Scholarship       110517
Hypertension      110517
Diabetes          110517
Alcoholism        110517
Handicap          110517
SMSSent           110517
NoShow            110517
dtype: int64

In [31]:
duplicateRowsDF = df[df.duplicated()]
duplicateRowsDF

Empty DataFrame
Columns: [PatientId, AppointmentID, Gender, ScheduledDay, AppointmentDay, Age, Neighborhood, Scholarship, Hypertension, Diabetes, Alcoholism, Handicap, SMSSent, NoShow]
Index: []

In [32]:
assert df.notnull().all().all()

### EXPORT DATA

In [33]:
df.to_csv('data/step2_output.csv', index = False)